File: HeesooPark/2022-09-20-FindMinimumInRotatedSortedArray.js
/**
 * @param {number[]} nums
 * @return {number}
 */
 var findMin = function(nums) {

    // should be O(log n) -> binary search

    // set the two pointers at 0 and n-1
    // set the guess m in the middle of the range
    let n = nums.length
    let l = 0
    let r = n-1
    let min = nums[0];


    // if nums[m] >= nums[l]
    // m is the part of the l's subarray
    // search right
    // if nums[m] < nums[l]
    // m is the part of the r's subarray
    // search left

    while ( l <= r) {

        if (nums[l] < nums[r]) {
            min = Math.min(min, nums[l])
            break
        }

        let m = Math.floor((l+r) /2)
        console.log(m)
        min = Math.min(min, nums[m])

        if (nums[m] >= nums[l]) {
            l = m + 1;
        } else {
            r = m - 1;
        }
    }

    return min;
}


// ### SOLUTION 2
function findMin(nums) {
    let l = 0;
    let r = nums.length - 1
    let m;

    while (l < r) {
        m = (l + r) / 2;

        if (nums[m] > nums[r]) {
            l = m + 1;
        } else {
            r = m;
        }
    }

    return nums[l]
}

File: BenKim/2022-09-20-find_minimum_in_rotated_sorted_array.py
class Solution:
    def findMin(self, nums: List[int]) -> int:
        
        # O(log n)의 시간복잡도를 갖기 위해서는 매 연산마다 앞으로의 연산수가 반으로 줄어야 한다
        # 정렬되어있는 배열이기 때문에, middle값과 right값의 크기 비교를 통해 최소값이 middle값의 좌우 어디에 있을지 예측할수 있다

        left = 0
        right = len(nums) - 1

        while left < right:
            middle = (left + right) // 2

            # 가운데수 뒤로 오름차순인경우(최소값이 middle이하의 인덱스에 있다)
            if nums[middle] < nums[right]:
                right = middle

            # 가운데수 뒤로 오름차순이 아닌 경우(최소값이 middle 초과의 인덱스에 있다)
            elif nums[middle] > nums[right]:
                left = middle + 1
        return nums[left]